lucene/sandbox/src/java/org/apache/lucene/search/LargeNumHitsTopDocsCollector.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.lucene.search;
+
+import java.io.IOException;
+
+import org.apache.lucene.index.LeafReaderContext;
+
+/**
+ * Optimized collector for large number of hits. This collector does
+ * not prepopulate the priority queue with sentinel values to avoid
+ * undue costs
+ */
+public class LargeNumHitsTopDocsCollector extends TopScoreDocCollector {
+
+  LargeNumHitsTopDocsCollector(int numHits, int totalHitsThreshold) {
```

Comment:
users won't be able to use it if it doesn't have a public constructor
also I don't think we need to pass a totalHitsThreshold: the fact that lots of hits need to be fetched would disable optimizations to the retrieval of top hits anyway

Reply:
+1, will fix